File: minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int largestValueProcessed = INT_MAX;

    int getMinimumDifference(TreeNode* root) {
        if (root == nullptr) return INT_MAX;

        int diff = getMinimumDifference(root->left);
        // no need for abs, this node will become our largest value so it's > largestValueProcessed
        if (largestValueProcessed != INT_MAX) diff = min(diff, root->val - largestValueProcessed);
        // now we can update our largest value since it would be "this node"
        largestValueProcessed = root->val;

        return min(getMinimumDifference(root->right), diff);
    }
};

File: minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    int? largestValueProcessed = null;
    
    // Since this is sorted we can rely on the fact that the tree will be increasing
    // and in decreasing order so we can just compare the difference between our node
    // and our children (recursively) and know that it will be the absolute minimum distance
    // because left < parent < right
    public int GetMinimumDifference(TreeNode root) {
        if (root == null) return int.MaxValue;

        // no need for abs, this node will become our largest value so it's > largestValueProcessed
        int diff = GetMinimumDifference(root.left);
        if (largestValueProcessed.HasValue) diff = Math.Min(diff, root.val - largestValueProcessed.Value);
        // now we can update our largest value since it would be "this node"
        largestValueProcessed = root.val;

        return Math.Min(GetMinimumDifference(root.right), diff);
    }
}